### DISCUSSION TITLE
--hls-duration and ad filtering

### DISCUSSION THREAD

**Discussion by aditorens** [streamlink/streamlink#5100] [Q&A] [answered]
Hello.
I need to record first 1-2 seconds of twitch-stream. I use --twitch-disable-ads and --hls-duration 00:00:02 for this.
But if stream have preroll-ads, my recording is stopping and there's no any file (because ads is longer ther 1-2 seconds and streamlink can't record anything at all).
It looks like this:

[plugins.twitch][info] Will skip ad segments
[plugins.twitch][info] Waiting for pre-roll ads to finish, be patient
[stream.hls][info] Stopping stream early after 5
[stream.hls][info] Filtering out segments and pausing stream output
[cli][error] Try 1/1: Could not open stream <TwitchHLSStream ['hls', 'https://video-weaver.a...

So. It there a way to record N seconds of twitch-stream but after (!) preroll-ads?
Thanks in advance.

BTW. 
I don't have a turbo-subscription, so my OAUTH-token is not dropping the ads from records (when I using --twitch-api-header "Authorization=OAuth 1234", I mean).

**Comment by bastimeyer** [comment, marked as answer]
`--hls-duration` in combination with ad filtering (or segment filtering in general) is unsupported. Twitch ads also mess with the HLS media sequence numbers, and a workaround hasn't been implemented for that, because it'd require rewriting some of the HLSStreamWorker logic.

**Reply by aditorens** [reply]
Ok, I get it.
Thanks.

Change my script. I forgot that I can use streamlink-python-api. This api can filter preroll-ads and read-write first N-megabytes of stream. That's fine for me.

**Reply by bastimeyer** [reply]
You can achieve the same thing by piping the output into `dd`:
```sh
streamlink --stdout URL best | dd bs=1M count=2 iflag=fullblock > recording.ts
```

**Reply by aditorens** [reply]
Wow. That's even better. Thanks a lot.

**Reply by bastimeyer** [reply]
Keep in mind though that limiting the file size results in a broken transport stream. `--hls-duration` on the other hand cuts off the stream at certain HLS segments, so the output is guaranteed to be valid.

----

There appears to be a race condition with the output stream when the pipe gets closed by `dd`. I can only reliably reproduce this on my notebook. It's not super important and won't affect the output stream, but it should be fixed at some point. The new `StreamRunner` implementation is not the culprit.
```
2097152 bytes (2.1 MB, 2.0 MiB) copied, 1.91302 s, 1.1 MB/s
[cli][info] Stream ended
error: Error when writing to output: [Errno 32] Broken pipe, exiting
[cli][info] Closing currently open stream...
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
BrokenPipeError: [Errno 32] Broken pipe
```

**Reply by aditorens** [reply]
Actually I need to write a few seconds/few MB of stream to make a screenshot from it.
I tested your algoithm and updated it with ffmpeg command. And it worked fine.
Something like this:

```
scr_command = f'streamlink --twitch-disable-ads --stdout {twitch_url} best | dd bs=1M count=2 iflag=fullblock | ffmpeg -y -i pipe: -vframes 1 -q:v 3 {jpg_file}'
scr_command_ret = subprocess.run(scr_command, capture_output=True, shell=True)
```

I need screenshot for my discord-notification.
Yes, I know that twitch have a jpg-preview-URL in API. But this image is blank for 2-3 minutes (after stream begins).
Maybe you can find better way :)
